WoodStar/WoodStar/TdsChannel.cs

```diff
@@ -0,0 +1,156 @@
+            header.WriteToBuffer(preloginBuffer);
+            preloginStream.WriteToBuffer(preloginBuffer);
+
+            await _stream.WriteAsync(preloginBuffer, 0, header.Length);
+            await _stream.FlushAsync();
+
+            ArrayPool<byte>.Shared.Return(preloginBuffer);
+
+            var preloginResponse = await ReadNextPacketAsync(ResponseType.Prelogin);
+            if (preloginResponse is PreloginPacket preloginPacket)
+            {
+                // Assign version etc info to connection
+            }
+        }
+
+        public async Task Login(string username, string password, string database)
+        {
+            var login7Stream = new Login7Stream
+            {
+                UserName = username,
+                Password = password,
+                Database = database,
+            };
+            var header = new TdsHeader(PacketType.Login, PacketStatus.EOM, login7Stream.Length + TdsHeader.HeaderSize, 0, 1);
+            var loginBuffer = ArrayPool<byte>.Shared.Rent(header.Length);
+            header.WriteToBuffer(loginBuffer);
+            login7Stream.WriteToBuffer(loginBuffer);
+
+            await _sslStream.WriteAsync(loginBuffer, 0, header.Length);
+            await _sslStream.FlushAsync();
+
+            ArrayPool<byte>.Shared.Return(loginBuffer);
+
+            await ReadNextPacketAsync(ResponseType.Login);
+        }
+
+        public async Task ExecuteSqlAsync(string sql)
+        {
+            var sqlBatchStream = new SqlBatchStream(
+                new StreamHeader
+                {
+                    TransactionDescriptor = new TransactionDescriptorHeader(0, 1)
+                },
+                sql);
+
+            var header = new TdsHeader(PacketType.SQLBatch, PacketStatus.EOM, sqlBatchStream.Length + TdsHeader.HeaderSize, 0, 1);
+            var sqlBatchBuffer = ArrayPool<byte>.Shared.Rent(header.Length);
+            header.WriteToBuffer(sqlBatchBuffer);
+            sqlBatchStream.WriteToBuffer(sqlBatchBuffer);
+            var b = string.Join(",", sqlBatchBuffer);
+            var c = string.Join(" ", sqlBatchBuffer.Select(e => e.ToString("X2")));
+
+            await _sslStream.WriteAsync(sqlBatchBuffer, 0, header.Length);
+            await _sslStream.FlushAsync();
+
+            ArrayPool<byte>.Shared.Return(sqlBatchBuffer);
+
+            await ReadNextPacketAsync(ResponseType.SqlBatchResult);
+        }
+
+        private async ValueTask<TdsPacket?> ReadNextPacketAsync(ResponseType expectedResponse)
+        {
+            while (true)
+            {
+                var reader = expectedResponse == ResponseType.Prelogin || expectedResponse == ResponseType.Login
+                    ? _pipeReader
+                    : _sslPipeReader;
+                var result = await reader.ReadAtLeastAsync(8);
+                var buffer = result.Buffer;
+
+                if (buffer.Length >= TdsHeader.HeaderSize)
```

Comment:
My idea was that:
1. `ReadAtLeastAsync` by default reads 8 bytes, but whenever we fail at `if (buffer.Length >= header.Length)`, `ReadAtLeastAsync` should read the size of the packet (`header.Length`).
2. Where is no point in checking `buffer.Length`, since `ReadAtLeastAsync` will always return at the very least 8 bytes. One exception is whenever a connection is closed, but that case can be handled by making a pipe throw on 0 bytes.

Reply:
I changed method to use ReadAtLeastAsync (twice first for header and then for packet size) rather than a loop. No error handling but now we don't have repeated checks over looping.

Reply:
Looks great, just a small nit: replace 8 in `_pipeReader.ReadAtLeastAsync(8)` with `TdsHeader.HeaderSize`.